RavenDB-12376: RvnMemoryMapPager - Posix 64Bit Part

aviviadi opened the pull request.

aviviadi committed:
RavenDB-12376: RvmMemoryMappedPager introduced, currently instead of PosixMemoryMappedPager only. [C# part]

aviviadi committed:
RavenDB-12376: RvmMemoryMappedPager introduced, currently instead of PosixMemoryMappedPager only. [C89 part + refactoring]

aviviadi committed:
RavenDB-12376: fix make.sh

aviviadi committed:
RavenDB-12376: librvnpal.linux.x64.so - with the latest

aviviadi committed:
RavenDB-12376: remove PosixMemoryMapPager

ayende commented on src/Voron/Impl/Paging/RvnMemoryMapPager.cs:
```diff
@@ -0,0 +1,214 @@
+            _logger = LoggingSource.Instance.GetLogger<StorageEnvironment>($"Pager-{file}");
+
+            if (initialFileSize.HasValue == false || initialFileSize.Value == 0) 
+                initialFileSize = SysInfo.PageSize * 16;
```
Validate here that the `initial page size` is a multiple of the page size and at least 64KB

ayende commented on src/Voron/Impl/Paging/RvnMemoryMapPager.cs:
```diff
@@ -0,0 +1,214 @@
+            var state = pagerState ?? _pagerState;
+
+            // ReSharper disable once InvertIf
+            if (SysInfo.CanPrefetch == PalDefinitions.True)
```
Why isn't this a bool?

ayende commented on src/Voron/Impl/Paging/RvnMemoryMapPager.cs:
```diff
@@ -0,0 +1,214 @@
+            {
+                try
+                {
+                    PalHelper.ThrowLastError(errorCode, $"rvn_create_and_mmap64_file failed on {(FAIL_CODES)rc}");
```
`FAIL_CODES` -> `PalFailCodes`, or something like that

ayende commented on src/Voron/Impl/Paging/RvnMemoryMapPager.cs:
```diff
@@ -0,0 +1,214 @@
+            var rc = rvn_create_and_mmap64_file(
+                file.FullPath,
+                initialFileSize.Value,
+                _copyOnWriteMode ? MMAP_OPTIONS.CopyOnWrite : MMAP_OPTIONS.None,
```
`MMAP_OPTIONS` -> `PalMmapFlags`

ayende commented on src/Voron/Impl/Paging/RvnMemoryMapPager.cs:
```diff
@@ -0,0 +1,214 @@
+            if (SysInfo.CanPrefetch == PalDefinitions.True)
+            {
+                // ReSharper disable once UnusedVariable
+                if (_pagerState.ShouldPrefetchSegment(pageNumber, out void* virtualAddress, out long bytes))
```
`out _` instead of unused variable

ayende commented on src/Voron/Impl/Paging/RvnMemoryMapPager.cs:
```diff
@@ -0,0 +1,214 @@
+                    foreach (var alloc in currentState.AllocationInfos)
+                    {
+                        metric.IncrementFileSize(alloc.Size);
+                        if (rvn_memory_sync(alloc.BaseAddress, alloc.Size, MSYNC_OPTIONS.MS_SYNC, out var errorCode) != 0)
```
`MSYNC_OPTIONS` -> `MemorySyncFlags`

ayende commented on src/Voron/Impl/Paging/RvnMemoryMapPager.cs:
```diff
@@ -0,0 +1,214 @@
+            base.ReleaseAllocationInfo(baseAddress, size);
+
+            var rc = rvn_unmap(baseAddress, size, DeleteOnClose ? PalDefinitions.True : PalDefinitions.False, out var errorCodeUnmap, out var errorCodeMadvise);
+            // TODO : is madvise don't need - should fail unmap call ? (currently we fail it)
```
I don't understand this TODO

aviviadi replied:
unmap exec madvise(DONT_NEED) and then unmap.  madvise might fail and unmap succeed.
should we fail on that call, should we log it, or should we ignore it? (ignore sounds right to me)

ayende replied:
No, when we want to `unmap`, that is what we care about.
If this is a temp file and the `madvice` failed, that is a missed optimization, not something we care about

aviviadi committed:
RavenDB-12376: Support RvnMemoryMapPager on mac osx-x64 (C code, librvnpal.dylib and verification)

ayende commented on src/Voron/Impl/Paging/RvnMemoryMapPager.cs:
```diff
@@ -0,0 +1,214 @@
+                file.FullPath,
+                initialFileSize.Value,
+                _copyOnWriteMode ? MMAP_OPTIONS.CopyOnWrite : MMAP_OPTIONS.None,
+                out _handle,
```
We need to implement a `SafeHandle` here

aviviadi replied:
Is it ok to leave the SafeHandle implementation to the Rvn MMP's Windows part?

ayende replied:
No, `SafeHandle` is something that we want for all platforms.
Basically, this is just a way to ensure that if we leak a handle, the GC's finalizer will dispose it.

ayende replied:
Basically, same as `SafeFileHandle`

aviviadi replied:
I ment - can we delay the implementation to the window's part PR. and then when we do that - also for posix.

ayende replied:
Why, it is a pretty simple one

ayende commented on src/Voron/Impl/Paging/RvnMemoryMapPager.cs:
```diff
@@ -0,0 +1,214 @@
+
+        protected override void DisposeInternal()
+        {
+            var rc = rvn_dispose_handle(FileName.FullPath, _handle, DeleteOnClose ? PalDefinitions.True : PalDefinitions.False, out var errorCodeClose, out var errorCodeUnlink);
```
Let do that through the `SafeHandle`, just make sure that we aren't throwing on dispose

ayende commented on src/Voron/Impl/Paging/RvnMemoryMapPager.cs:
```diff
@@ -0,0 +1,214 @@
+
+        protected internal override void PrefetchRanges(Win32MemoryMapNativeMethods.WIN32_MEMORY_RANGE_ENTRY* list, int count)
+        {
+            // TODO : Get rid of WIN32_MEMORY_RANGE_ENTRY and use Pal's PrefetchRanges instead
```
Yes, very much so

aviviadi replied:
Will be done when we implement other pagers (added to [TODOs](https://issues.hibernatingrhinos.com/issue/RavenDB-12376#focus=streamItem-67-33365-0-0))

ayende commented on src/Voron/Impl/Paging/RvnMemoryMapPager.cs:
```diff
@@ -0,0 +1,214 @@
+        {
+            // TODO : Get rid of WIN32_MEMORY_RANGE_ENTRY and use Pal's PrefetchRanges instead
+            // ReSharper disable once UnusedVariable
+            rvn_prefetch_ranges((PalDefinitions.PrefetchRanges*)list, count, out var errorCode);
```
`out _`

ayende commented on src/Voron/Impl/Paging/RvnMemoryMapPager.cs:
```diff
@@ -0,0 +1,214 @@
+            var rc = rvn_allocate_more_space(newLength, _totalAllocationSize, FileName.FullPath, _handle, _copyOnWriteMode ? MMAP_OPTIONS.CopyOnWrite : MMAP_OPTIONS.None,
+                out var newAddress, out var newLengthAfterAdjustment, out var errorCode);
+
+            if (rc == (int)FAIL_CODES.FAIL_ALLOCATION_NO_RESIZE)
```
Why is this not an error?

aviviadi replied:
It is a flag to return null, in order to be consistent with previous implementation (like [here](https://github.com/ravendb/ravendb/blob/v4.1/src/Voron/Impl/Paging/AbstractPager.cs#L315))

ayende replied:
But this isn't actually an error. This is good, we don't need to do anything because we are big enough.
For that matter, can we move this check to the managed code?

aviviadi replied:
if we move this back to managed, we need : _nearest_size_to_page_size.  It will not require pinvoke as we already have the page size from get_system_information.  
Do you want to re-implement _nearest_size_to_page_size in managed ?

ayende replied:
We already have the page size in managed code, so there is nothing else we need.
We can do that in managed code, yes.
Still need to _verify_ that the desired size is page aligned.

ayende commented on src/Voron/Impl/Paging/RvnMemoryMapPager.cs:
```diff
@@ -0,0 +1,214 @@
+
+            if (rc != 0)
+            {
+                // TODO : can't allocate more pages, should we force throw out of mem (as was done before RvnMemoryMapPager introduced ?
```
This depends on what the failure was.
Did we fail to allocate more disk space ( `DiskFull` or fail to actually mmap the memory `OOM`)

aviviadi replied:
So, we are left with the native error code handling `GetNativeErrorString(lastError, msg, out var specialErrnoCodes)` which can return NoMem and NoSpc.  In posix we rely on the native error code returned by the syscalls. in windows we will return NoMem or NoSpc according to exceptions I presume

ayende replied:
What we get back (always) is out own codes.
If they happen to coincide in Linux with something, great.
But I would worry about that across distros and would rather use our own directly always

ayende commented on src/Voron/Impl/Paging/RvnMemoryMapPager.cs:
```diff
@@ -0,0 +1,214 @@
+                UsePageProtection == false && force == false) 
+                return;
+
+            if (rvn_protect_range(start, (long)size, MPROTECT_OPTIONS.PROT_READ, out var errorCode) == 0)
```
`MPROTECT_OPTIONS` -> `MemoryProtectionFlags`

ayende commented on src/Voron/Platform/Pal.cs:
```diff
@@ -1,24 +1,22 @@
-// ReSharper disable SwitchStatementMissingSomeCases
-// ReSharper disable InconsistentNaming
+using static Voron.Platform.PalDefinitions;
+// ReSharper disable BuiltInTypeReferenceStyle
```
??

ayende commented on src/Voron/Platform/Pal.cs:
```diff
@@ -75,7 +73,74 @@ static Pal()
+            string filename,
+            Int64 initialFileSize,
+            PalFlags.MMAP_OPTIONS flags,
+            out void* handle,
```
Need to write our own `SafeHandle` impl to take care of this.

ayende commented on src/Voron/Platform/Pal.cs:
```diff
@@ -75,7 +73,74 @@ static Pal()
+            string filepath,
+            void* handle,
+            Int32 deleteOnClose,
+            out Int32 errorCodeClose,
```
Only one error code should be returned, you can check the return value to check what was the actual error.

aviviadi replied:
but we might have them both fail and we want to log both of them (to understand later if and why we have leftovers)

ayende replied:
That doesn't matter, we fail early and return the earlier failure code.
Note that we still want to _close_ the file, but if we are already in error mode, we don't care if that failed.

ayende replied:
The reasoning behind this is that we can't expose our internal state and behavior outside. That changes over time.
What we want to do is to have a common error handling interface.
Return code, `int32_t` with our own error code.
`detailed_error_code` with the system error code.

Any additional data you want, must be encoded in the return code, nothing else.

ayende commented on src/Voron/Platform/Pal.cs:
```diff
@@ -75,7 +73,74 @@ static Pal()
+            void* address,
+            Int64 size,
+            Int32 deleteOnClose,
+            out Int32 errorCodeUnmap,
```
Only one error code should be returned, you can check the return value to check what was the actual error.

aviviadi replied:
here it makes sense to leave one error code if madvise error should be ignored and maybe even not returning madvise error at all if we are not logging it. Is this still the case? why not leaving the option to read all the failures? (if it is a signature issue, we can implement `(Int32 *)errorCodeArray`)

ayende replied:
Because you are baking the implementation behavior directly into the interface.
What if we have a new behavior, we want to not have to change the signature.
For example, if we decided that encrypted files need to be zeroed first, and that fails, do we now need _another_  error code?

That violates design guidelines and something we want to avoid.

ayende commented on src/Voron/Platform/PalDefinitions.cs:
```diff
@@ -0,0 +1,26 @@
+    public unsafe class PalDefinitions
+    {
+        public static Int32 True => 1;
+        public static Int32 False => 0;
```
Kill these two, there is no such thing, we use `bool`

ayende commented on src/Voron/Platform/PalDefinitions.cs:
```diff
@@ -0,0 +1,26 @@
+        public struct SystemInformation
+        {
+            public Int32 PageSize;
+            public Int32 CanPrefetch;
```
```suggestion
            public bool CanPrefetch;
```

aviviadi replied:
As suggested from [this](https://stackoverflow.com/questions/20035826/why-dllimport-for-c-bool-as-unmanagedtype-i1-throws-but-as-byte-it-works?lq=1) and [this](https://docs.microsoft.com/en-us/cpp/dotnet/calling-native-functions-from-managed-code?view=vs-2017) C# `bool` must be marshaled (in our case to Int32, or maybe we want to change it to c89's byte?)

aviviadi replied:
I ended up with `[MarshalAs(UnmanagedType.I4)]` for the bool type to make sure we write int32_t into that bool and not something else (as it might change between OSs and compilers)

ayende replied:
`I4` is fine, yes

aviviadi replied:
I4 doesn't work so I am left with U1

ayende replied:
The problem is that the moment we use `char` in here, that mess up alignment, padding, etc.
We want to ensure that this is not going to cause us (very subtle) problems down the line, so we want to make the compiler's job as simple as possible.
That means that we have to be strict about alignment, help the compiler avoid the need for padding, etc.

aviviadi replied:
so we have to cast from bool to int on the pInvoke calling

ayende replied:
Or use `int`

aviviadi replied:
I originally used Int32.. get back to it ?

ayende replied:
yes

ayende commented on src/Voron/Platform/PalDefinitions.cs:
```diff
@@ -0,0 +1,26 @@
+        public static Int32 True => 1;
+        public static Int32 False => 0;
+
+        [SuppressMessage("ReSharper", "UnassignedField.Global")]
```
Why unused?

ayende commented on src/Voron/Platform/PalDefinitions.cs:
```diff
@@ -0,0 +1,26 @@
+            public Int32 CanPrefetch;
+        }
+
+        [SuppressMessage("ReSharper", "UnusedMember.Global")]
```
Why unused?

ayende commented on src/Voron/Platform/PalFlags.cs:
```diff
@@ -1,13 +1,13 @@
 // ReSharper disable IdentifierTypo
+// ReSharper disable InconsistentNaming
+// ReSharper disable UnusedMember.Global
+
```
I really don't like the R# notifications, let's fix them

ayende commented on src/Voron/Platform/PalFlags.cs:
```diff
@@ -1,13 +1,13 @@
     public static class PalFlags
     {
-        [Flags]
         public enum FAIL_CODES : uint
```
```suggestion
        public enum FailureFlags : uint
```

ayende replied:
Also, this probably should be `int`, not `uint`

ayende commented on src/Voron/Platform/PalFlags.cs:
```diff
@@ -21,14 +21,50 @@ public enum FAIL_CODES : uint
             FAIL_RACE_RETRIES = 7,
             FAIL_PATH_RECURSION = 8,
             FAIL_FLUSH_FILE = 9,
+            FAIL_SYSCONF = 10,
```
Use C# naming conventions, not UPPER_CASE

ayende replied:
This is about all of the stuff in this file

ayende commented on src/Voron/Platform/PalFlags.cs:
```diff
@@ -21,14 +21,50 @@ public enum FAIL_CODES : uint
+        }
+
+        [Flags]
+        public enum MSYNC_OPTIONS
```
This is a bad idea, you are reflecting the OS options to higher level code.
Instead, let's say what we _want_.
And unless I miss something, we always pass the same flags.

ayende commented on src/Voron/Platform/PalFlags.cs:
```diff
@@ -21,14 +21,50 @@ public enum FAIL_CODES : uint
+        }
+
+        [Flags]
+        public enum MPROTECT_OPTIONS
```
Same, we don't need to expose these flags, we need to say what we want.
`MemoryProtectionFlags` with `ReadOnly`, `ReadWrite` are the only things we use

ayende commented on src/Voron/Platform/PalHelper.cs:
```diff
@@ -1,30 +1,34 @@
                 txt = $"Errno: {lastError}='{nativeMsg}' (rc={specialErrnoCodes}) - '{msg}'";
 
-                if ((specialErrnoCodes & (int)PalFlags.ERRNO_SPECIAL_CODES.ENOMEM) != 0)
+                if ((specialErrnoCodes & (int)PalFlags.ERRNO_SPECIAL_CODES.ENOMEM) != 0 || forceThrowOutOfMemory)
```
```suggestion
                if (((PalFlags)specialErrnoCodes & PalFlags.ERRNO_SPECIAL_CODES.ENOMEM) != 0 || forceThrowOutOfMemory)
```

ayende commented on src/Voron/Platform/PalHelper.cs:
```diff
@@ -1,30 +1,34 @@
                 if ((specialErrnoCodes & (int)PalFlags.ERRNO_SPECIAL_CODES.ENOENT) != 0)
                     throw new FileNotFoundException(txt);
+
+                if ((specialErrnoCodes & (int)PalFlags.ERRNO_SPECIAL_CODES.ENOSPC) != 0)
```
```suggestion
                if (((PalFlags)specialErrnoCodes & PalFlags.ERRNO_SPECIAL_CODES.ENOSPC) != 0)
```

ayende commented on src/Voron/Platform/Posix/PosixHelper.cs:
```diff
@@ -27,7 +27,7 @@ public static void AllocateFileSpace(StorageEnvironmentOptions options, int fd,
             {
                 var diskSpaceResult = DiskSpaceChecker.GetDiskSpaceInfo(file);
-                throw new DiskFullException(file, size, diskSpaceResult?.TotalFreeSpace.GetValue(SizeUnit.Bytes));
+                throw new DiskFullException(file, size, diskSpaceResult?.TotalFreeSpace.GetValue(SizeUnit.Bytes), result.ToString());
```
Why are you calling `ToString` on the `result`? That wouldn't give a good error message

ayende commented on src/Voron/Platform/Win32/Win32NativeFileMethods.cs:
```diff
@@ -128,7 +128,7 @@ public static void SetFileLength(SafeFileHandle fileHandle, long length)
                 {
                     var driveInfo = DiskSpaceChecker.GetDiskSpaceInfo(filePath);
-                    throw new DiskFullException(filePath, length, driveInfo?.TotalFreeSpace.GetValue(SizeUnit.Bytes));
+                    throw new DiskFullException(filePath, length, driveInfo?.TotalFreeSpace.GetValue(SizeUnit.Bytes), lastError.ToString());
```
Same, `lastError` is a string, we need an actual error

ayende commented on src/librvnpal/inc/rvn.h:
```diff
@@ -4,34 +4,88 @@
+EXPORT struct SYSTEM_INFORMATION
+{
+    int32_t page_size;
+    int32_t can_prefetch;
```
```suggestion
    bool can_prefetch;
```

ayende commented on src/librvnpal/inc/rvn.h:
```diff
@@ -4,34 +4,88 @@
-int32_t /* internal use */
-sync_directory_allowed (int dir_fd);
+EXPORT int32_t
+rvn_dispose_handle(const char *filepath, void *handle, int32_t delete_on_close, int32_t *unlink_error_code, int32_t *close_error_code);
```
Should have just one `int32_t` error. We can't expose our internal impl in this manner

ayende commented on src/librvnpal/inc/rvn.h:
```diff
@@ -4,34 +4,88 @@
+rvn_dispose_handle(const char *filepath, void *handle, int32_t delete_on_close, int32_t *unlink_error_code, int32_t *close_error_code);
 
+EXPORT int32_t
+rvn_unmap(void *address, int64_t size, int32_t delete_on_close, int32_t *unmap_error_code, int32_t *madvise_error_code);
```
One error out var

ayende commented on src/librvnpal/inc/rvn.h:
```diff
@@ -4,34 +4,88 @@
+
+/* For internal use: */
+PRIVATE int64_t
+rvn_nearest_size_to_page_size(int64_t orig_size, int64_t sys_page_size);
```
`rvn_` prefix is for public stuff, let's use a different one

ayende commented on src/librvnpal/inc/status_codes.h:
```diff
@@ -2,26 +2,49 @@
+#define MMOPTIONS_COPY_ON_WRITE         1
+
+#define MSYNC_OPTIONS_MS_SYNC           (1 << 0)
+#define MSYNC_OPTIONS_MS_ASYNC          (1 << 1)
```
When do we use this?

ayende commented on src/librvnpal/inc/status_codes.h:
```diff
@@ -2,26 +2,49 @@
+#define MPROTECT_OPTIONS_PROT_NONE      0
+#define MPROTECT_OPTIONS_PROT_READ      (1 << 0)
+#define MPROTECT_OPTIONS_PROT_WRITE     (1 << 1)
+#define MPROTECT_OPTIONS_PROT_EXEC      (1 << 2)
```
When do we use this?

I really want to avoid copying the structure of the OS flags

ayende commented on src/librvnpal/src/posix/fileutils.c:
```diff
@@ -0,0 +1,136 @@
+int64_t
+rvn_pwrite(int32_t fd, void *buffer, uint64_t count, uint64_t offset, int32_t *detailed_error_code)
+{
+    int64_t actually_written = 0;
```
```suggestion
    uint64_t actually_written = 0;
```

ayende commented on src/librvnpal/src/posix/fileutils.c:
```diff
@@ -0,0 +1,136 @@
+    int64_t cifs_retries = 3;
+    do
+    {
+        int64_t result = pwrite(fd, buffer, count - (uint64_t)actually_written, offset + actually_written);
```
```suggestion
        int64_t result = pwrite(fd, buffer, count - actually_written, offset + actually_written);
```

ayende commented on src/librvnpal/src/posix/fileutils.c:
```diff
@@ -0,0 +1,136 @@
+                /* cifs/nfs mount can sometimes fail on EINVAL after file creation
+                lets give it few retries with short pauses between them - RavenDB-11954 */
+
+                sleep(1);
```
Use `usleep` and the shorter sleep times

ayende commented on src/librvnpal/src/posix/fileutils.c:
```diff
@@ -0,0 +1,136 @@
+}
+
+int32_t
+rvn_pointer_to_int(void *ptr)
```
Why is this required? we can just explicitly cast to `int32_t`, no?
I believe that this could should be a macro with:

```
#define PtrToInt(p) ((int32_t)(intptr_t)p))

```

aviviadi replied:
Simple cast generates compiler warning which want to avoid just here

ayende replied:
```//gcc 5.4.0

#include  <stdio.h>


int main(void)
{
    
    void * a = (void*)5;
    int  b = (int)(long)a;
    printf("Hello, world! %i", b);
    return 0;
}```

Compiles with now warnings. Note the _two_ casts, we are being explicit about it

ayende commented on src/librvnpal/src/posix/fileutils.c:
```diff
@@ -0,0 +1,136 @@
+}
+
+int32_t
+rvn_dispose_handle(const char *filepath, void *handle, int32_t delete_on_close, int32_t *unlink_error_code, int32_t *close_error_code)
```
Single out error

ayende commented on src/librvnpal/src/posix/geterrorstring.c:
```diff
@@ -18,8 +23,11 @@ int32_t rvn_get_error_string(int32_t error, char* buf, int32_t buf_size, int32_t
+			break;
 		default:
-			*special_errno_flags = ERRNO_SPECIAL_CODES_NONE;
+			*special_errno_flags = ERRNO_SPECIAL_CODES_NO_ERROR;
```
`SUCCESS` ?

aviviadi replied:
NONE instead of SUCCESS (zero value). it is not Success necessarily

ayende commented on src/librvnpal/src/posix/mapping.c:
```diff
@@ -0,0 +1,174 @@
+
+    assert(initial_file_size > 0);
+
+    /* TODO: rnv_ensure_file_exists(path); */
```
isn't this done by `O_CREAT` ?

aviviadi replied:
no, it is not.

ayende commented on src/librvnpal/src/posix/mapping.c:
```diff
@@ -0,0 +1,174 @@
+
+    if (sz <= initial_file_size || sz % sys_page_size != 0)
+    {
+        sz = rvn_nearest_size_to_page_size(max(initial_file_size, sz), sys_page_size);
```
The min here isn't `sys_page_size` but `allocation granularity`, `4KB` on linux, 64Kb on windows, let's always make it 64KB

ayende commented on src/librvnpal/inc/rvn.h:
```diff
@@ -4,34 +4,88 @@
+rvn_get_error_string(int32_t error, char *buf, int32_t buf_size, int32_t *special_errno_flags);
+
+EXPORT int32_t
+rvn_create_and_mmap64_file(const char *path, int64_t initial_file_size, int32_t flags, void **handle, void **base_addr, int64_t *actual_file_size, int32_t *detailed_error_code);
```
Let's document here what this means, that we are getting read/write memory, etc. 
so the flags that we are actually using and their impact

ayende commented on src/librvnpal/src/posix/fileutils.c:
```diff
@@ -0,0 +1,136 @@
+        }
+    }
+
+    int32_t close_rc = close(fd);
```
Check that fd != -1 ?

ayende commented on src/librvnpal/src/posix/mapping.c:
```diff
@@ -0,0 +1,174 @@
+}
+
+int32_t
+rvn_allocate_more_space(int64_t new_length, int64_t total_allocation_size, const char *filename, void *handle, int32_t flags,
```
I don't understand the signature here.
You are operating of a file, so that should be first. 
what is the new_legnth and total allocation for?

aviviadi replied:
total_allocation_size is the "previous" size and this is for knowing if we need to increase the file size or return "AlloctionNoResize". 
The new_length.. is the new length to resize to..

ayende replied:
But we already know that in the managed code, so why call out to here anyway?

ayende commented on src/librvnpal/src/posix/mapping.c:
```diff
@@ -0,0 +1,174 @@
+
+    *new_length_after_adjustment = rvn_nearest_size_to_page_size(new_length, sys_page_size);
+
+    if (*new_length_after_adjustment <= total_allocation_size)
```
I assume that total_allocation_size is the current size?  Is this method increasing the size of the allocation?

aviviadi replied:
> I assume that total_allocation_size is the current size? Is this method increasing the size of the allocation?

it returns new_length_after_adjustment and the caller sets the total_allocation_size to the new value.

ayende commented on src/librvnpal/src/posix/mapping.c:
```diff
@@ -0,0 +1,174 @@
+}
+
+int32_t
+rvn_unmap(void *address, int64_t size, int32_t delete_on_close, int32_t *unmap_error_code, int32_t *madvise_error_code)
```
single out error

ayende commented on src/librvnpal/src/posix/sync.c:
```diff
@@ -10,134 +10,163 @@
+rvn_memory_sync(void *address, int64_t size, int32_t flags, int32_t *detailed_error_code)
+{
+    int32_t msync_flags = 0;
+    if (flags & MSYNC_OPTIONS_MS_SYNC)
```
Nope, we aren't going to expose the same interface as the OS has, we need to expose our own, much simpler interface

ayende commented on src/librvnpal/src/posix/virtualmemory.c:
```diff
@@ -0,0 +1,68 @@
+        {
+            rc = prefetch_rc;
+            *detailed_error_code = prefetch_errno;
+        }
```
On any failure, we immediately return

ayende commented on src/librvnpal/src/posix/virtualmemory.c:
```diff
@@ -0,0 +1,68 @@
+    for (i = 0; i < count; i++)
+    {
+        struct RVN_RANGE_LIST record = range_list[i];
+        int32_t prefetch_rc = rvn_prefetch_virtual_memory(record.virtual_address, record.number_of_bytes, &prefetch_errno);
```
```suggestion
        int32_t prefetch_rc = rvn_prefetch_virtual_memory(record.virtual_address, record.number_of_bytes, detailed_error_code);
```

aviviadi replied:
I want to record the first error but continue with the other ranges. The above suggestion will override the errno if second iteration successful.

ayende replied:
No, it is a bad idea. We want to exit early

ayende replied:
We got an error, and continuing work from there means that we don't know what is our state.
given that this is an optimization only, that doesn't matter

ayende commented on src/librvnpal/src/posix/virtualmemory.c:
```diff
@@ -0,0 +1,68 @@
+rvn_protect_range(void *start_address, int64_t size, int32_t flags, int32_t *detailed_error_code)
+{
+    int32_t mprotect_flags = 0;
+    if (flags & MPROTECT_OPTIONS_PROT_NONE)
```
Nope, we don't expose the OS state outside, only our own state / options

ayende commented on src/librvnpal/src/posix/writefileheader.c:
```diff
@@ -10,65 +10,72 @@
-  if (syncIsNeeded == true)
-    return sync_directory_for (path, detailed_error_code);
-  return SUCCESS;
+rvn_write_header(const char *path,
```
Does this include @haludi 's changes?

aviviadi replied:
nop.. @haludi will merge later on using this function

ayende commented on src/librvnpal/src/posix/writefileheader.c:
```diff
@@ -10,65 +10,72 @@
+
+    int32_t remaining = size;
+
+    /* do not call rvn_get_file_size to avoid 3 lseek calls, and use only 2 calls */
```
that isn't a meaningful optimization, we write this very rarely

aviviadi replied:
we don't have rvn_get_file_size anyhow..

ayende commented on src/librvnpal/src/posix/writefileheader.c:
```diff
@@ -10,65 +10,72 @@
+
+    while (remaining > 0)
+    {
+        uint64_t written = write(fd, header, (uint64_t)remaining);
```
Can just use `pwrite` here and avoid setting calling `lseek` again.
Although, can call `stat` here, which likely will be fastest option

ayende commented:
And, to verify, this doesn't yet include windows impl, right?

ayende commented on src/librvnpal/src/posix/mapping.c:
```diff
@@ -1,5 +1,11 @@
 #if defined(__unix__) || defined(__APPLE__)
 
+#ifdef __APPLE__
+#define MMAP mmap
```
We can probably just do `mmap64` ->`mmap` here, and avoid `MMAP`

aviviadi replied:
as a precocious i would like to leave it this way, as :
```
void *mmap    (void *addr, size_t len, int prot, int flags, int fildes, off_t off);
void *mmap64(void *addr, size_t len, int prot, int flags, int fildes, off64_t off); 
``` 
in [here](https://www.mkssoftware.com/docs/man3/mmap.3.asp) it states : `The mmap64() function is a part of the large file extensions. `

ayende replied:
so call it `rvn_mmap` instead, I don't like `MMAP`

ayende commented on src/librvnpal/src/posix/virtualmemory.c:
```diff
@@ -53,10 +53,12 @@ rvn_protect_range(void *start_address, int64_t size, int32_t flags, int32_t *det
         mprotect_flags |= PROT_WRITE;
     if (flags & MPROTECT_OPTIONS_PROT_EXEC)
         mprotect_flags |= PROT_EXEC;
+#ifndef __APPLE__        
```
All of this is not required. And this is exactly why not.
We don't need these options, and exposing them cause us problem

aviviadi committed:
RavenDB-12376: add librvnpal for arm32 with the above changes plus verification

aviviadi commented:
> And, to verify, this doesn't yet include windows impl, right?

no, it is not.

aviviadi committed:
RavenDB-12376: Fix comments (excluding TODOs from https://issues.hibernatingrhinos.com/issue/RavenDB-12376#focus=streamItem-67-33365-0-0). plus verified librvnpal.linux.x64.so only

aviviadi committed:
Fix check rc of sync allowed plus minors

aviviadi committed:
RavenDB-12376: support cross compilation for arm32

aviviadi committed:
Merge branch 'v4.2' of https://github.com/ravendb/ravendb into RavenDB-12376_posixMMP

aviviadi committed:
RavenDB-12376: fix remarks - single errorcode

aviviadi committed:
RavenDB-12376: fix remarks - allocate more space: calc page alignment in managed code

aviviadi committed:
RavenDB-12376: fix remarks - fix bool to int marshaling / usage

aviviadi committed:
RavenDB-12376: fix remarks - fix bool to int - now back to Int32 but this time with PalFlags

aviviadi committed:
RavenDB-12376
fix remarks (refactoring (FileCloseFlags etc), cast int32/64_t instead of int long, revert mistake on mac FULLFSYNC)
- not including corresponding libs yet

aviviadi committed:
RavenDB-12376: support osx-x64 (mac os) cross compilation

aviviadi committed:
RavenDB-12376: fix cross compilation for armhf plush remake all posix libs (linux-x64 + arm-32 verified, osx-64 dylib about to be verified)

aviviadi committed:
RavenDB-12376: add arm64 (aarch64) cross compilation support plus lib (not verified)

aviviadi commented:
will be continued at #8502
